src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderLabel.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Security.Cryptography.Cose
+{
+    [DebuggerDisplay("Label = {LabelAsInt32.HasValue ? LabelAsInt32 : LabelAsString}, Type = {LabelAsInt32.HasValue ? typeof(int) : typeof(string),nq}")]
+    public readonly struct CoseHeaderLabel : IEquatable<CoseHeaderLabel>
+    {
+        // https://www.iana.org/assignments/cose/cose.xhtml#header-parameters
+        public static readonly CoseHeaderLabel Algorithm = new CoseHeaderLabel(KnownHeaders.Alg);
+        public static readonly CoseHeaderLabel Critical = new CoseHeaderLabel(KnownHeaders.Crit);
+        public static readonly CoseHeaderLabel ContentType = new CoseHeaderLabel(KnownHeaders.ContentType);
+        public static readonly CoseHeaderLabel KeyIdentifier = new CoseHeaderLabel(KnownHeaders.Kid);
+        public static readonly CoseHeaderLabel IV = new CoseHeaderLabel(KnownHeaders.IV);
+        public static readonly CoseHeaderLabel PartialIV = new CoseHeaderLabel(KnownHeaders.PartialIV);
+        public static readonly CoseHeaderLabel CounterSignature = new CoseHeaderLabel(KnownHeaders.CounterSignature);
+
+        internal int? LabelAsInt32 { get; }
+        internal string? LabelAsString { get; }
+
+        public CoseHeaderLabel(int label)
+        {
+            LabelAsInt32 = label;
+            LabelAsString = null;
+        }
+
+        public CoseHeaderLabel(string label)
+        {
+            LabelAsInt32 = null;
+            LabelAsString = label ?? throw new ArgumentException(null, nameof(label));
+        }
+
+        public bool Equals(CoseHeaderLabel other) => LabelAsInt32.HasValue ? LabelAsInt32 == other.LabelAsInt32 : LabelAsInt32 == other.LabelAsInt32;
+
+        public override bool Equals(object? obj) => obj is CoseHeaderLabel otherObj && Equals(otherObj);
+
+        public override int GetHashCode() => base.GetHashCode();
```

Comment:
This should compute a hashcode rather than delegating to ValueType.GetHashCode.

Reply:
Would this be a good HashCode implementation?
```cs
public override int GetHashCode() => LabelAsInt32?.GetHashCode() ?? LabelAsInt32.GetHashCode();
```

Should I also overload the `==` and `!=` operators?

Reply:
> Would this be a good HashCode implementation?

I think so, but make sure that

```C#
CoseHeaderLabel label = default;
label.GetHashCode();
```

doesn't throw.  I think it'll be zero, but worth having a test for.

> Should I also overload the == and != operators?

The book says

> 8.13.1 Equality Operators on Value Types
> * DO overload the equality operators on value types, if equality is meaningful.

So, apparently :smile: